### PR TITLE
Add ADR implementation reference

### PR DESCRIPTION
**Pull request by benolayinka**
#### Summary

Closes https://github.com/TheThingsIndustries/lorawan-stack-docs/issues/274

#### Screenshots

Bildschirmfoto 2021-06-23 um 11 29 59

#### Changes

- Add ADR implementation reference

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Run Locally: Verified that the docs build using `make server`, posted screenshots, verified external links. Test with `HUGO_PARAMS_SEARCH_ENABLED=true` if style changes will affect the search bar.
- [ ] New Features Marked: Documentation for new features is marked using the `new-in-version` shortcode, according to the guidelines in [CONTRIBUTING](CONTRIBUTING.md).
- [x] Style Guidelines: Documentation obeys style guidelines in [CONTRIBUTING](CONTRIBUTING.md).
- [x] Commits: Commit messages follow guidelines in [CONTRIBUTING](CONTRIBUTING.md), there are no fixup commits left.

### CODE REVIEW

File: doc/content/reference/adr/_index.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+1. Determine [the maximum SNR over recent transmissions](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L218)
+2. Determine [the minimum SNR to demodulate an uplink given the current parameters](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L232)
+3. Calculate [the margin to further optimize the data rate](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L236)
+  - Part of this is configurable per device (if you use the CLI)
+  - If less measurements (uplinks) are available than necessary, [include a safety margin](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L238-L240)
+4. [Increase the data rate as long as there's enough margin](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L251-L262)
+5. If there's still margin after reaching the maximum data rate, [decrease the transmit power](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L273-L281)
```
Comment (nejraselimovic) [inline doc/content/reference/adr/_index.md:37]:
```suggestion
4. [Increase the data rate as long as there is enough margin](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L251-L262)
5. If there is still margin after reaching the maximum data rate, [decrease the transmit power](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L273-L281)
```

**Review by nejraselimovic — APPROVED** [review]

**Review by johanstokking — COMMENTED** [review]
Nice.

If you include line examples, please use a commit hash instead of `v3.13`, because those lines may shift then.

Looping in also @adriansmares to look at the contents.

I did a quick pass and it looks good to me.

File: doc/content/reference/adr/_index.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+## The Things Stack ADR Algorithm
+
+This section describes {{% tts %}} ADR implementation in plain english, with links to the relevant lines of source code in our open source [LoRaWAN Stack Repository](https://github.com/TheThingsNetwork/lorawan-stack).
+
+{{% tts %}} ADR implementation looks at the signal-to-noise ratio (SNR) to determine the Data Rate and Tx Power, and looks at frame counters to determine packet loss, and set the number of re-transmissions accordingly.
+
+The implementation is based on Semtech's recommended algorithm described in [this document](https://www.thethingsnetwork.org/forum/uploads/default/original/2X/7/7480e044aa93a54a910dab8ef0adfb5f515d14a1.pdf):
+
+1. Determine [the maximum SNR over recent transmissions](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L218)
```
Comment (adriansmares) [inline doc/content/reference/adr/_index.md:31]: https://github.com/TheThingsNetwork/lorawan-stack/blob/5a816e8171f993db9659566286d45725698f032e/pkg/networkserver/mac/adr.go#L218

File: doc/content/reference/adr/_index.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+2. Determine [the minimum SNR to demodulate an uplink given the current parameters](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L232)
```
Comment (adriansmares) [inline doc/content/reference/adr/_index.md:32]: https://github.com/TheThingsNetwork/lorawan-stack/blob/5a816e8171f993db9659566286d45725698f032e/pkg/networkserver/mac/adr.go#L232

File: doc/content/reference/adr/_index.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+3. Calculate [the margin to further optimize the data rate](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L236)
```
Comment (adriansmares) [inline doc/content/reference/adr/_index.md:33]: https://github.com/TheThingsNetwork/lorawan-stack/blob/5a816e8171f993db9659566286d45725698f032e/pkg/networkserver/mac/adr.go#L236

File: doc/content/reference/adr/_index.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  - Part of this is configurable per device (if you use the CLI)
+  - If less measurements (uplinks) are available than necessary, [include a safety margin](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L238-L240)
```
Comment (adriansmares) [inline doc/content/reference/adr/_index.md:35]: https://github.com/TheThingsNetwork/lorawan-stack/blob/5a816e8171f993db9659566286d45725698f032e/pkg/networkserver/mac/adr.go#L238-L240

File: doc/content/reference/adr/_index.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+4. [Increase the data rate as long as there's enough margin](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L251-L262)
```
Comment (adriansmares) [inline doc/content/reference/adr/_index.md:36]: https://github.com/TheThingsNetwork/lorawan-stack/blob/5a816e8171f993db9659566286d45725698f032e/pkg/networkserver/mac/adr.go#L251-L262

File: doc/content/reference/adr/_index.md (hunk 1 above)
Comment (adriansmares) [reply doc/content/reference/adr/_index.md:37]: https://github.com/TheThingsNetwork/lorawan-stack/blob/5a816e8171f993db9659566286d45725698f032e/pkg/networkserver/mac/adr.go#L273-L281

File: doc/content/reference/adr/_index.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+5. If there's still margin after reaching the maximum data rate, [decrease the transmit power](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L273-L281)
+6. Depending on packet loss, [increase the number of retransmissions](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.13/pkg/networkserver/mac/adr.go#L288-L296)
```
Comment (adriansmares) [inline doc/content/reference/adr/_index.md:38]: https://github.com/TheThingsNetwork/lorawan-stack/blob/5a816e8171f993db9659566286d45725698f032e/pkg/networkserver/mac/adr.go#L288-L296

**Review by adriansmares — APPROVED** [review]
This looks good. I've commented the permanent links where applicable.

**Review by johanstokking — APPROVED** [review]
Approved when using the permanent links that @adriansmares commented
